src/odemis/driver/smaract.py

```diff
@@ -3173,3 +3174,1239 @@ def SA_CTL_Move(self, handle, ch, pos_pm, _):
+        return ev
+
+    def _load_configuration(self):
+        """
+        Load the configuration of the device. The loaded parameters are
+        * fiber length
+        * extension fiber length
+        * number of active channels
+        * sensor head types
+        * minimum and maximum working distance
+        These configuration parameters need to be set during the installation of the system.
+        Additionally, the parameters from the previously saved referencing procedure are loaded.
+        This function is blocking.
+        """
+        # Loading the configuration will cause the state to become temporarily unstable. Once the process is
+        # finished, the state will be stable again.
+        self.EnableEventNotification(SA_PS_STABLE_STATE_CHANGED_EVENT)
+        self.SetProperty_i32(SA_PS_AF_ADJUSTMENT_RESULT_LOAD_PROP, SA_SI_ENABLED)
+        self._wait_for_progress_event(SA_SI_ENABLED, timeout=600)  # this can take a long time
+        self.DisableEventNotification(SA_PS_STABLE_STATE_CHANGED_EVENT)
+
+    def _save_configuration(self):
+        """
+        Saves the configuration parameters (cf _load_configuration) and internal parameters of the
+        referencing procedure. This function is blocking.
+        """
+        self.SetProperty_i32(SA_PS_AF_ADJUSTMENT_RESULT_SAVE_PROP, SA_SI_ENABLED)
+        # There is no documentation on loading/saving the configuration in the manual, even though
+        # the functions are available in the header file. It is not clear how much time saving the
+        # parameters takes. It can be assumed that it's fast --> wait 1 s.
+        time.sleep(1)
+
+    def _wait_for_progress_event(self, end_state, timeout=600):
+        """
+        Blocks until progress event is triggered or timeout.
+        It is assumed that only one event type is enabled. The function will wait until the
+        .devEventParameter contains the end_state. This can mean different things depending on the event.
+        end_state (int): state of event variable to wait for
+        timeout (float): maximum time to wait in s. If inf, it will wait forever.
+        """
+        if timeout == float("inf"):
+            tend = None
+        else:
+            tend = time.time() + timeout
+        ev = SA_SI_Event()
+
+        state = None
+        while state != end_state:
+            if state is not None:
+                logging.debug("Skipped event 0x%x" % ev.type)
+
+            if tend is not None:
+                t = tend - time.time()
+                if t < 0:
+                    raise TimeoutError("Timeout limit of %s s exceeded." % timeout)
+            else:
+                t = float("inf")
+            ev = self.WaitForEvent(t)
+            # Don't check event type, it's not always accurate.
+            # Instead, only one event is active and all other events are disabled, so it is clear
+            # which event we are waiting for.
+            state = ev.devEventParameter & 0xffff
```

Comment:
You *have* to check the event type. The rest of the content of `ev` has completely different meaning depending on the event type. So "state" has only a meaning when the even type is for progress event.
I'd suggest to write something like:
```python
ev = self.WaitForEvent(t)  # raise TimeOut/CancelledError by itself
if ev.type == SA_PS_AF_ADJUSTMENT_PROGRESS_EVENT:
      # lowest 16-bits is the adjustment state
      state = ev.devEventParameter & 0xffff
elif ev.type == SA_SI_STREAM_ABORTED_EVENT:
    raise SA_SIError(ev.error, ...)
else:
    logging.debug("Skipped event 0x%x" % ev.type)
```

Reply:
I also use the function to check the STABLE_STATE_CHANGED event. Most events seem to work in the same way (cf Table 7, p 223). I think I tried to use the type in the beginning, but then it turned out that it wasn't always correct with all events. Now I'm making much less use of the events, so maybe it will work, but I need to check on the hardware. If it doesn't I should probably report it to Smaract.